chanfitness/chaneventstore.go

```diff
@@ -0,0 +1,381 @@
+// Package chanfitness monitors the behaviour of channels to provide scores
+// which can be factored into the decision to close a channel. This is
+// achieved by maintaining an event store which tracks events for each
+// channel.
+//
+// Lifespan: the period that the channel has been known to the scoring system.
+// Note that lifespan may not equal the channel's full lifetime because data is
+// not currently persisted.
+//
+// Uptime: the total time within a given period that the channel's remote peer
+// has been online.
+package chanfitness
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/channelnotifier"
+	"github.com/lightningnetwork/lnd/peernotifier"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// ChannelEventStore maintains a set of event logs for the node's channels.
+// It is intended to provide insight into the performance of channels.
+type ChannelEventStore struct {
+	cfg *Config
+
+	// channels maps short channel IDs to event logs.
+	channels map[uint64]*chanEventLog
+
+	// peers tracks the current online status of peers based on online/offline
+	// events.
+	peers map[[33]byte]bool
+
+	// lifespanRequests serves requests for the lifespan of channels.
+	lifespanRequests chan lifespanRequest
+
+	// uptimeRequests serves requests for the uptime of channels.
+	uptimeRequests chan uptimeRequest
+
+	quit chan struct{}
+
+	wg sync.WaitGroup
+}
+
+// Config provides the event store with functions required to monitor channel
+// activity. All elements of the config must be non-nil for the event store to
+// operate.
+type Config struct {
+	// Channel events provides a subscription client which provides a
+	// stream of channel events.
+	ChannelEvents func() (*subscribe.Client, error)
```

Comment:
Rename this to `SubscribeChannelEvent`, to make it clear that every invocation creates a new client?